cpp/src/arrow/compute/kernels/vector_selection.cc

```diff
@@ -2378,26 +2378,44 @@ struct NonZeroVisitor {
     using T = typename GetViewType<Type>::T;
     uint32_t index = 0;
 
-    VisitArrayDataInline<Type>(
-        this->array,
-        [&](T v) {
-          if (v) {
-            this->builder->UnsafeAppend(index);
-          }
-          ++index;
-        },
-        [&]() { ++index; });
+    for (ArrayDataVector::const_iterator current_array = arrays.begin();
+         current_array != arrays.end(); ++current_array) {
+      VisitArrayDataInline<Type>(
+          **current_array,
+          [&](T v) {
+            if (v) {
+              this->builder->UnsafeAppend(index);
+            }
+            ++index;
+          },
+          [&]() { ++index; });
+    }
+
     return Status::OK();
   }
 };
 
 Status IndicesNonZeroExec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
-  std::shared_ptr<ArrayData> array = batch[0].array();
   UInt64Builder builder;
+  ArrayDataVector arrays;
+  Datum input = batch[0];
+
+  if (input.kind() == Datum::ARRAY) {
+    std::shared_ptr<ArrayData> array = input.array();
+    RETURN_NOT_OK(builder.Reserve(array->length));
+    arrays.push_back(array);
```

Comment:
nit `arrays.push_back(std::move(array))`

Reply:
:+1: as the shared_ptrs pushed into the array are destroy once the block is exited, there was no need to increase references and as you mentioned we could easily steal the existing one.